Update to Compose 1.5.0-beta02

malliaridis opened the pull request:
This is the continuation of #441 (recreated PR due to rebase to `v2.1` branch).

This PR should contain only the necessary version updates as discussed in #441.

malliaridis committed:
Update dependencies to support Compose 1.5.0-beta01

Update several library versions in deps.versions.toml to support
Compose 1.5.0-beta01.

Additional changes in `Pages` and `*.gradle.kts` files are necessary
due to breaking changes in `ComposeCompilerKotlinSupportPlugin`,
`rememberPagerState`, `HorizontalPager` and `VerticalPager`.

IMPORTANT: This commit introduces breaking changes for `Pages`.

arkivanov commented on sample/shared/dynamic-features/feature1Impl/build.gradle.kts:
```diff
@@ -65,11 +65,11 @@ compose.web.targets()
 plugins.removeAll { it is ComposeCompilerKotlinSupportPlugin }
 
-class ComposeNoNativePlugin : KotlinCompilerPluginSupportPlugin by ComposeCompilerKotlinSupportPlugin() {
+class ComposeNoNativePlugin : KotlinCompilerPluginSupportPlugin by ComposeCompilerKotlinSupportPlugin({}) {
```
Can we use named arguments for the lamda, here and below?

malliaridis replied:
Sure thing. I hope the conventions / formatting match the one you use in the project. :)

arkivanov commented on sample/shared/dynamic-features/feature2Impl/build.gradle.kts:
```diff
@@ -65,11 +65,11 @@ compose.web.targets()
 plugins.removeAll { it is ComposeCompilerKotlinSupportPlugin }
 
-class ComposeNoNativePlugin : KotlinCompilerPluginSupportPlugin by ComposeCompilerKotlinSupportPlugin() {
+class ComposeNoNativePlugin : KotlinCompilerPluginSupportPlugin by ComposeCompilerKotlinSupportPlugin({}) {
```
Same here.

arkivanov reviewed (changes requested):
Could you please also update API files? Please run `./gradlew apiDump` and commit updated files.

malliaridis committed:
Add argument names to ComposeCompilerKotlinSupportPlugin

arkivanov commented:
> Could you please also update API files? Please run `./gradlew apiDump` and commit updated files.

Could you please also update API files? The build fails otherwise.

malliaridis commented:
> > Could you please also update API files? Please run `./gradlew apiDump` and commit updated files.
> 
> Could you please also update API files? The build fails otherwise.

Sorry, I missed that one. As [mentioned in the previous PR](https://github.com/arkivanov/Decompose/pull/441#discussion_r1285711005), Kotlin 1.9.0. is required in order to proceed, which makes the current version of the `gradle-setup-plugin` incompatible (it seems to fail due to the `android.kotlinOptions` from [here](https://github.com/arkivanov/gradle-setup-plugin/blob/0f3baf84e9dc32118e5d46141fb3eec901831a9a/src/main/java/com/arkivanov/gradle/SetupAndroid.kt#L57) that couldn't be resolved).

@arkivanov Should I create another PR for the setup-plugin, or do you prefer to postpone this change for now?

arkivanov commented:
Updating to Kotlin 1.9.0 is a bit complicated, we also need to update parcelize-darwin-plugin, and then Essenty. Even though it may be working fine with updates, still good to update all together.

Let me try updating those libs, I'll update here.

arkivanov commented:
Meantime, would it be possible to update API files, or is it not compiling?

malliaridis commented:
Calling `./gradlew apiDump` fails with 

```
FAILURE: Build failed with an exception.

* Where:
Build file '[PATH_TO_PROJECT]/decompose/build.gradle.kts' line: 17

* What went wrong:
Could not find method kotlinOptions() for arguments [org.gradle.kotlin.dsl.GroovyBuilderScope$closureFor$1@78405471] on extension 'android' of type com.android.build.gradle.LibraryExtension.
> Could not find method kotlinOptions() for arguments [org.gradle.kotlin.dsl.GroovyBuilderScope$closureFor$1@78405471] on extension 'android' of type com.android.build.gradle.LibraryExtension.
```

EDIT: After that there are no changes to commit. So it looks like it fails before updating the API files.
EDIT2: I did update the kotlin version to 1.9.0 before `apiDump` because it was failing prior with `This version (1.5.0) of the Compose Compiler requires Kotlin version 1.9.0`. With Kotlin 1.8.20 I get changes in the API files, but I am not sure if I should commit those before updating Kotlin.

arkivanov commented:
Ok, thanks! Let's wait for 1.9.0 update then. I will update here.

arkivanov commented:
Essenty is updated to Kotlin 1.9.0, however it was pretty difficult to update. Could you please merge `v2.1` into this branch, and also do the same changes in `extensions-compose-jetpack` module ([here](https://github.com/arkivanov/Decompose/blob/v2.1/extensions-compose-jetpack/src/main/java/com/arkivanov/decompose/extensions/compose/jetpack/pages/Pages.kt))? After that, I will push a commit with Kotlin 1.9.0 update into this branch.

arkivanov commented on deps.versions.toml:
```diff
@@ -6,12 +6,12 @@ essenty = "1.2.0-alpha-03"
 jetbrainsBinaryCompatibilityValidator = "0.12.1"
-jetpackCompose = "1.4.0"
-jetpackComposeCompiler = "1.4.5"
+jetpackCompose = "1.5.0-beta03"
```
Perhaps we can now use Jetpack Compose `1.5.0`?

malliaridis replied:
Jetbrains Compose `1.5.0-beta02` makes use of Compose (Runtime, UI and more)`1.5.0`, so I updated both.

arkivanov commented on deps.versions.toml:
```diff
@@ -6,12 +6,12 @@ essenty = "1.2.0-alpha-03"
 reaktive = "1.2.3"
 junit = "4.13.2"
-jetbrainsCompose = "1.4.0"
+jetbrainsCompose = "1.5.0-beta01"
```
Compose Multiplatform [1.5.0-beta02](https://github.com/JetBrains/compose-multiplatform/releases/tag/v1.5.0-beta02) is available. Maybe use that one?

malliaridis committed:
Merge remote-tracking branch 'origin/v2.1' into compose-1.5.0-beta01

malliaridis committed:
Update Pages from Jetpack Compose extension

Remove `pageCount` from extension to follow the API changes from Compose.

malliaridis committed:
Update JetBrains Compose to 1.5.0-beta02

Update jetbrainsCompose to `1.5.0-beta02` and
the corresponding compatibility version of
jetpackCompose to `1.5.0`.

IMPORTANT: Additional dependencies, including Kotlin,
need to be updated in order to run the current state.

malliaridis commented:
> Essenty is updated to Kotlin 1.9.0, however it was pretty difficult to update. Could you please merge v2.1 into this branch, and also do the same changes in extensions-compose-jetpack module ([here](https://github.com/arkivanov/Decompose/blob/v2.1/extensions-compose-jetpack/src/main/java/com/arkivanov/decompose/extensions/compose/jetpack/pages/Pages.kt))? After that, I will push a commit with Kotlin 1.9.0 update into this branch.

I updated the Pages from jetpack compose extension now too. Good catch by the way.

You can start with the update to Kotlin `1.9.0`. I assume AGP, `targetSdk` etc. need to be updated too? I haven't executed any `apiDump` yet because of the (currently) incompatible Kotlin version.

arkivanov committed:
Updated Kotlin to 1.9.0, AGP to 8.0.2, Gradle to 8.0.2, parcelize-darwin to 0.1.5

arkivanov commented on extensions-compose-jetbrains/src/commonMain/kotlin/com/arkivanov/decompose/extensions/compose/jetbrains/pages/Pages.kt:
```diff
@@ -52,10 +53,13 @@ fun <T : Any> Pages(
-    val state = rememberPagerState(initialPage = selectedIndex)
+    val state = rememberPagerState(
+        initialPage = selectedIndex,
+        pageCount = { pages.items.size },
```
After reviewing the code further, it looks like the remembered `pageCount` lambda captures the instance of `ChildPages`, which is immutable. So even if a new instance of `ChildPages` of another size will be be supplied later, the lambda will be still capturing the old one.

It feels that this overload doesn't make any sense now, and we can only have the one with `Value<ChildPages>`. WDYT?

arkivanov replied:
Or perhaps even better, let the second overload accept `State<ChildPages>` instead of just `ChildPages`. And make sure that `pageCount` lambda reads the actual value from the provided `State`.

```kotlin
@Composable
fun <T : Any> Pages(
    pages: State<ChildPages<*, T>>,
    onPageSelected: (index: Int) -> Unit,
    modifier: Modifier = Modifier,
    scrollAnimation: PagesScrollAnimation = PagesScrollAnimation.Disabled,
    pager: Pager = defaultHorizontalPager(),
    pageContent: @Composable PagerScope.(index: Int, page: T) -> Unit,
)
```

What's your opinion on this?

malliaridis replied:
Now that I have an updated and working version I can see the problem too. I think your suggested alternative with `State<ChildPages<*, T>>` would be more flexible for the users, since they have more options for using `Pages` (e.g. when using `StateFlow` instead of `Value`).

Should I update the second function so that it consumes the state passed via `pages` and use the newly consumed state instead? And is it an issue if consuming twice the state in the case of `Pages` with `pages: Value`?

malliaridis replied:
In addition to that, can we create a few tests for the updating part of `ChildPages`? Because I am experiencing a confusing `IndexOutOfBoundsException` when I simply use a mutable state property and update the pages from two to three pages.

I was trying to add the following test for the variant `pages: State`. I feel like I am missing something crucial:

```kotlin
@OptIn(ExperimentalFoundationApi::class, ExperimentalDecomposeApi::class)
@Test
fun GIVEN_pages_dislpayed_WHEN_pages_state_updated_THEN_pageCount_updated() {
    runBlocking(Dispatchers.Main) {
        val state = mutableStateOf(
            ChildPages(
                items = listOf(
                    Child.Created(Config.Config1, Unit),
                    Child.Destroyed(Config.Config2),
                ),
                selectedIndex = 0,
            ),
        )

        composeRule.setContent {
            Pages(
                pages = state,
                onPageSelected = {},
            ) { index, page ->
                Text("$index $page")
            }
        }

        composeRule.awaitIdle()
        state.value = ChildPages(
            items = listOf(
                Child.Destroyed(Config.Config1),
                Child.Created(Config.Config2, Unit),
                Child.Destroyed(Config.Config3), // <-- Tried both, created and destroyed
            ),
            selectedIndex = 1,
        )
        composeRule.waitForIdle() // <-- Fails here with IndexOutOfBoundsException
    }
}

private sealed class Config {

    object Config1: Config()

    object Config2: Config()

    object Config3: Config()
}
```

arkivanov replied:
Thanks! Tests are always welcome!

I have reproduced that crash without Decompose, filed https://issuetracker.google.com/issues/295653367 just in case. Let's see how it goes.

I believe that's a Compose bug, but here is a code that seems to work.

```kotlin
@ExperimentalFoundationApi
@ExperimentalDecomposeApi
@Composable
fun <T : Any> Pages(
    pages: State<ChildPages<*, T>>,
    onPageSelected: (index: Int) -> Unit,
    modifier: Modifier = Modifier,
    scrollAnimation: PagesScrollAnimation = PagesScrollAnimation.Disabled,
    pager: Pager = defaultHorizontalPager(),
    pageContent: @Composable PagerScope.(index: Int, page: T) -> Unit,
) {
    val selectedIndex = pages.value.selectedIndex
    val state = rememberPagerState(
        initialPage = selectedIndex,
        pageCount = { pages.value.items.size },
    )

    LaunchedEffect(selectedIndex) {
        if (state.currentPage != selectedIndex) {
            when (scrollAnimation) {
                is PagesScrollAnimation.Disabled -> state.scrollToPage(selectedIndex)
                is PagesScrollAnimation.Default -> state.animateScrollToPage(page = selectedIndex)
                is PagesScrollAnimation.Custom -> state.animateScrollToPage(page = selectedIndex, animationSpec = scrollAnimation.spec)
            }
        }
    }

    DisposableEffect(state.settledPage) {
        onPageSelected(state.settledPage)
        onDispose {}
    }

    pager(
        modifier,
        state,
        { pages.value.items[it].configuration },
    ) { pageIndex ->
        pages.value.items[pageIndex].instance?.also { page ->
            pageContent(pageIndex, page)
        }
    }
}
```

What do you think?

malliaridis replied:
Seems like when the `pager` receives an immutable value through the function it doesn't update properly, like with the `rememberPagerState.pageCount` for which we introduced the `State<ChildPages<*, T>>`.

The solution you provided works now with the test too. I will Add a commit with the updates later today and look into adding a few tests afterwards. :)

malliaridis replied:
While adding tests I noticed that they are not in `commonTest`. Is there a specific reason for that or can I add the `PagesTests` to `commonTest`?

arkivanov replied:
Tests are in `jvmMain` because they depend on Compose for Desktop testing stuff. I think we can't run common tests for Compose yet?

arkivanov replied:
Seems like tests were not committed. Do you still want to add the tests?

malliaridis replied:
I haven't committed them yet because I wasn't sure if I should add them into `common` code. But if they are not yet supported I will have to add them for `jvm` only.

arkivanov replied:
`jvmTest` seems like a good place for now. Thanks!

malliaridis committed:
Fix Pages state update issues

Update the Pages state observation to
properly update on state changes.

arkivanov commented:
Seems like API files have to be updated once again. Could you please run `./gradlew apiDump` and commit the files?

malliaridis committed:
Add positive tests for Pages

Add positive tests for state changes in Pages.

malliaridis committed:
API dump

malliaridis committed:
Add state changes tests for Pages (Android)

arkivanov commented:
Thanks for you contribution!

azlekov commented:
Hey, not sure where to ask but Jetpack compose stable version is already 1.5.1 
@arkivanov is there anything waiting for stable Decompose 2.1.0? Thanks!

arkivanov commented:
@azlekov Decompose [2.1.0-beta-01](https://github.com/arkivanov/Decompose/releases/tag/2.1.0-beta-01) is based on Kotlin 1.9.0 and Multiplatform Compose 1.5.0, which should be binary compatible with Compose 1.5.1. The update to Compose 1.5.1 has been merged in #475, but not yet released. The stable version 2.1.0 is blocked by a unavailability of a stable version of [androidx.activity](https://developer.android.com/jetpack/androidx/releases/activity) (currently the latest version is `1.8.0-beta01`. I could release Decompose `2.1.0-beta-02` with Multiplatform Compose 1.5.1, but I believe you should be fine with `2.1.0-beta-01`.

azlekov commented:
> @azlekov Decompose [2.1.0-beta-01](https://github.com/arkivanov/Decompose/releases/tag/2.1.0-beta-01) is based on Kotlin 1.9.0 and Multiplatform Compose 1.5.0, which should be binary compatible with Compose 1.5.1. The update to Compose 1.5.1 has been merged in #475, but not yet released. The stable version 2.1.0 is blocked by a unavailability of a stable version of [androidx.activity](https://developer.android.com/jetpack/androidx/releases/activity) (currently the latest version is `1.8.0-beta01`. I could release Decompose `2.1.0-beta-02` with Multiplatform Compose 1.5.1, but I believe you should be fine with `2.1.0-beta-01`.

Thanks for the info @arkivanov!
